sdk/src/platform/Discovery.ts

```diff
@@ -0,0 +1,362 @@
+        this._refreshHandicapMs = refreshHandicapMs;
+        this._refreshDelayMs = refreshDelayMs;
+        this._initialEndpoint = initialEndpoint;
+        this._fetchGet = fetchGet;
+        this._clientId = clientId;
+
+        this._initialRetryMaxCount = retryCount;
+        this._initialRetryInterval = retryInterval;
+
+        this._externalRetryMaxCount = retryCount;
+        this._externalRetryInterval = retryInterval;
+
+        this.init();
+    }
+
+    public async init() {
+        if (!this._clientId) {
+            throw new Error('Client Id is required for discovery');
+        }
+        if (!this._initialPromise) {
+            this._initialPromise = this._init();
+        }
+        try {
+            await this._initialPromise;
+            this._initialPromise = null;
+        } catch (e) {
+            this._initialPromise = null;
+            throw e;
+        }
+    }
+
+    private async _init() {
+        let initialData = await this.initialData();
+        if (!initialData) {
+            initialData = await this.fetchInitialData();
+        } else {
+            this._initialRetryMaxCount = initialData.retryCount;
+            this._initialRetryInterval = initialData.retryInterval;
+            this._externalRetryMaxCount = initialData.retryCount;
+            this._externalRetryInterval = initialData.retryInterval;
+        }
+        this._initialized = true;
+        this.emit(events.initialized, initialData);
+    }
+
+    public async fetchInitialData() {
+        if (!this._initialFetchPromise) {
+            this._initialFetchPromise = this._fetchInitialData();
+        }
+        try {
+            const initialData = await this._initialFetchPromise;
+            this._initialRetryMaxCount = initialData.retryCount;
+            this._initialRetryInterval = initialData.retryInterval;
+            this._externalRetryMaxCount = initialData.retryCount;
+            this._externalRetryInterval = initialData.retryInterval;
+            this._initialFetchPromise = null;
+            return initialData;
+        } catch (e) {
+            this._initialFetchPromise = null;
+            this.emit(events.initialFetchError, e);
+            throw e;
+        }
+    }
+
+    private async _fetchInitialData() {
+        try {
+            const response = await this._fetchGet(
+                this._initialEndpoint,
+                {clientId: this._clientId},
+                {skipAuthCheck: true, skipDiscoveryCheck: true},
+            );
+            const initialData = await response.json();
+            await this._setInitialData(initialData);
+            this._initialRetryCount = 0;
+            return initialData;
+        } catch (e) {
+            this._initialRetryCount += 1;
+            if (this._initialRetryCount < this._initialRetryMaxCount) {
+                await delay(this._initialRetryInterval * 1000);
+                return this._fetchInitialData();
```

Comment:
Do we want to retry on ALL errors? What if not network related errors will occur?

Reply:
Good point. we should not retry for 401